lib/ecto/query/builder.ex

```diff
@@ -267,6 +267,12 @@ defmodule Ecto.Query.Builder do
   def escape(nil, _type, params_acc, _vars, _env),
     do: {nil, params_acc}
 
+  def escape(atom, type, {params, acc}, vars, _env) when is_atom(atom) do
+    expr = literal(atom, type, vars)
+    params = [{atom, type} | params]
+    {expr, {params, acc}}
+  end
```

Comment:
Hi @sabiwara!

I don't think we should convert it to a parameter, because parameters are passed as interpolation. I believe you can rewrite this code as:

```suggestion
  def escape(atom, type, params_acc, vars, _env) when is_atom(atom) do
    {atom, params_acc}
  end
```

Have you tried this?

Reply:
Hi @josevalim, thanks a lot for the review!

Yes this was actually my first attempt, but I wasn't able to have the following case raise without adding it as a param:
https://github.com/elixir-ecto/ecto/blob/6dfc2f512e462f292f84542ed5a4d49648f15534/test/ecto/query/planner_test.exs#L262

From what I understood, the `Builder` stage isn't able to detect a type mismatch yet, because it is unaware of the schema?
So I tried to understand how it was done for other similar errors, and ended up making it a param, so that `Planner` can check it when running `cast_and_merge_params`.

It might have been kind of a hack, telling Ecto to "treat `:foo` as you would have treated `^:foo`".
But this was exactly the behavior I had in mind: always cast atoms.

Any pointer for where to look without using params? I would like to to keep trying a bit if it is OK :slightly_smiling_face:

Reply:
If that approach doesn't work, then I don't know *why* it doesn't work, so I need to jump in and take a look. :) The fix may be simple but finding it may be much harder. :D